src/main/scala/org/constellation/domain/healthcheck/MissingProposalHealthCheckConsensusManager.scala

```diff
@@ -0,0 +1,273 @@
+
+  override def onSuccessfullRoundStart(): F[Unit] = F.unit
+
+  override def readyConsensusesAction(
+    consensuses: Map[MissingProposalHealthCheckKey, HealthCheckConsensus[
+      F,
+      MissingProposalHealthCheckKey,
+      MissingProposalHealthCheckStatus,
+      MissingProposalHealthStatus,
+      SendPerceivedHealthStatus[
+        MissingProposalHealthCheckKey,
+        MissingProposalHealthCheckStatus,
+        MissingProposalHealthStatus
+      ]
+    ]]
+  ): F[Unit] = F.unit
+
+  override def positiveOutcomeAction(
+    positiveOutcomePeers: Map[
+      MissingProposalHealthCheckKey,
+      (
+        HealthcheckConsensusDecision[MissingProposalHealthCheckKey],
+        HealthCheckConsensus[
+          F,
+          MissingProposalHealthCheckKey,
+          MissingProposalHealthCheckStatus,
+          MissingProposalHealthStatus,
+          SendPerceivedHealthStatus[
+            MissingProposalHealthCheckKey,
+            MissingProposalHealthCheckStatus,
+            MissingProposalHealthStatus
+          ]
+        ]
+      )
+    ]
+  ): F[Unit] =
+    positiveOutcomePeers.toList.traverse {
+      case (key, (_, consensus)) =>
+        for {
+          ownStatus <- consensus.getOwnConsensusHealthStatus().map(_.status)
+          _ <- if (ownStatus.isInstanceOf[GotPeerProposalAtHeight])
+            logger.debug(s"Already have the proposal for key=${logHealthCheckKey(key)}. Omitting proposal download!")
+          else
+            retrieveMissingProposal(key, consensus).handleErrorWith(
+              e =>
+                logger
+                  .warn(e)(
+                    s"Failed while retrieving missing proposal at height=${key.height} for id=${logIdShort(key.id)}"
+                  )
+            )
+        } yield ()
+    }.void
+
+  private def retrieveMissingProposal(
+    key: MissingProposalHealthCheckKey,
+    consensus: HealthCheckConsensus[
+      F,
+      MissingProposalHealthCheckKey,
+      MissingProposalHealthCheckStatus,
+      MissingProposalHealthStatus,
+      SendPerceivedHealthStatus[
+        MissingProposalHealthCheckKey,
+        MissingProposalHealthCheckStatus,
+        MissingProposalHealthStatus
+      ]
+    ]
+  ): F[Unit] =
+    for {
+      _ <- logger.debug(s"Retrieving missing proposal for key=${logHealthCheckKey(key)}")
+      roundPeers <- consensus.getRoundPeers()
+      peersWithMissingProposal = roundPeers.collect {
+        case (id, RoundData(_, Some(MissingProposalHealthStatus(_, _, _, _, GotPeerProposalAtHeight(_, _), _)), _)) =>
+          id
+      }.toList.take(1)
+      // TODO: take all peers from above and try until first success instead of taking only one
+      _ <- logger.debug(
+        s"Found following peers=${logIds(peersWithMissingProposal.toSet)} that may have the missing proposal key=${logHealthCheckKey(key)}"
+      )
+      _ <- peersWithMissingProposal.traverse(
+        redownloadService.fetchProposalForPeer(key.id)(_)
```

Comment:
Wait, I think we shouldn't download missing proposals from a health check. There is already a mechanism for that.

Reply:
Yeah, we could do nothing when the healthcheck outcome is positive and wait for the redownloadService to find the gap, or we could try to download the proposal here also as a reaction to a positive healthcheck outcome.

Reply:
I think we should keep these responsibilities split. Downloading missing proposals should happen in a single place so that we have control over when it happens and how it happens, i.e. only when we're stuck and only from a short-range of heights.